Kafka 14565:  Interceptor Resource Leak

beardt opened the pull request:
Added try/catch within Abstract::getConfigInstances to capture interceptor configuration exceptions and enable clean up of resources used by previous created interceptor instances listed within the interceptor.classes property .

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

beardt committed:
Added try/catch with resource clean up to getConfiguredInstances

beardt committed:
Added unit tests for AbstractConfig.getConfiguredInstances memory leak fix.  Also, included Closeable instance check in the try/catch.

C0urante commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -476,14 +480,34 @@ public <T> List<T> getConfiguredInstances(List<String> classNames, Class<T> t, M
+        } catch (Exception e) {
+            for (Object object : objects) {
+                if (object instanceof AutoCloseable) {
+                    try {
+                        ((AutoCloseable) object).close();
+                    } catch (Exception ex) {
+                        log.error(String.format("Close exception on %s", object.getClass().getName()), ex);
+                    }
+                } else if (object instanceof Closeable) {
+                    try {
+                        ((Closeable) object).close();
+                    } catch (Exception ex) {
+                        log.error(String.format("Close exception on %s", object.getClass().getName()), ex);
+                    }
```
This can be simplified a bit:

```suggestion
                    Utils.closeQuietly((AutoCloseable) object, "AutoCloseable object constructed and configured during failed call to getConfiguredInstances");

```

1. `Utils::closeQuietly` handles failures for us
2. `Closeable` is a subinterface of `AutoCloseable`, so we only need to check for the latter

beardt replied:
@C0urante  I'm glad to hear that you are feeling better.  
Yes, given the KIP avoidance constraint, I really did not want to change getConfiguredInstances, but  it seemed the only way to address this was to change it very carefully.   However,  I believe I've implemented each of your comments and thanks for the excellent feedback.

C0urante commented on clients/src/test/java/org/apache/kafka/test/MockConsumerInterceptor.java:
```diff
@@ -55,6 +58,11 @@ public void configure(Map<String, ?> configs) {
+
+        CONFIG_COUNT.incrementAndGet();
+        if (CONFIG_COUNT.get() == THROW_CONFIG_EXCEPTION_THRESHOLD.get()) {
+            throw new ConfigException("Kafka producer creation failed. Failure may not have cleaned up listener thread resource.");
```
It seems like the failure message here is hinting that we try to create a Kafka producer in this interceptor, but there isn't much else in the class to go along with that.

Could we use a more generic message like "Failed to instantiate interceptor (reached throw-on-config threshold)"?

beardt replied:
Done!

C0urante commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -503,6 +505,30 @@ public void testInterceptorConstructorClose() {
     }
 
+    @Test
+    public void testInterceptorConstructorConfigurationWithExceptionShouldCloseRemainingInstances() {
```
I like this test. Could we also get a matching one for producer interceptors, and something analogous for the `AbstractConfig` class?

C0urante reviewed (commented):
Thanks @beardt, and sorry for the delay (was out sick last week). I really like this approach!

I have to admit it's not quite what I had in mind since it does change the `AbstractConfig` class, but the more I think about it, the more this feels like a complete, proper fix rather than some kind of workaround, and I can't envision a reasonable scenario where this might accidentally break downstream code.

That said, it'd be nice to get another committer's opinion on this to make sure we don't have to go through a KIP for this change. @ijuma @guozhangwang would either of you have time to weigh in here?

beardt committed:
- Simplified AbstractConfig::getConfiguredInstances try/catch logic to use Utils::closeQuietly.
- Added unit test KafkaProducerTest.
- Modified configuration threshold exception message to be more generic.
- Add unit test to AbstractConfigTest.

beardt committed:
- Cleaned up unit test variable naming convention and mock interceptor constants in AbstractConfigTest, KafkaProducerTest and KafkaConsumerTest.

C0urante commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -476,14 +479,22 @@ public <T> List<T> getConfiguredInstances(List<String> classNames, Class<T> t, M
+            }
+        } catch (Exception e) {
+            for (Object object : objects) {
+                Utils.closeQuietly((AutoCloseable) object, "AutoCloseable object constructed and configured during failed call to getConfiguredInstances");
```
We still need an `instanceof` check here:
```suggestion
                if (object instanceof AutoCloseable) {
                    Utils.closeQuietly((AutoCloseable) object, "AutoCloseable object constructed and configured during failed call to getConfiguredInstances");
                }
```

C0urante replied:
Actually, come to think of it, we might also want to invoke `close` on objects created in `getConfiguredInstance` if they throw an exception from `configure` [here](https://github.com/apache/kafka/blob/6c98544a964b40ede6bbe1b3440f8e5db96a4ad6/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L407).

Perhaps we could pull this out into a reusable method and use it both here and there? Thinking something like:

```java
private static void maybeClose(Object object, String name) {
    if (object instanceof AutoCloseable)
        Utils.closeQuietly(object, name);
}
```

C0urante commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -137,6 +137,7 @@
 import static org.mockito.Mockito.when;
 
 public class KafkaProducerTest {
+    private final int targetInterceptor = 3;
```
Nit: If we're not using this field in any other test, probably makes more sense to inline directly into the test case it's used in.

C0urante commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -168,6 +169,7 @@ public class KafkaConsumerTest {
     private final int defaultApiTimeoutMs = 60000;
     private final int requestTimeoutMs = defaultApiTimeoutMs / 2;
     private final int heartbeatIntervalMs = 1000;
+    private final int targetInterceptor = 3;
```
Nit: If we're not using this field in any other test, probably makes more sense to inline directly into the test case it's used in.

beardt replied:
Just an observation, I see the following variable is used in only one test case  `private final int throttleMs = 10;` as well.

C0urante commented on clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:
```diff
@@ -45,6 +47,8 @@
 
 public class AbstractConfigTest {
 
+
+
```
We don't need this change; it's fine as-is.

C0urante commented on clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:
```diff
@@ -54,6 +58,12 @@ public void testConfiguredInstances() {
         testInvalidInputs("org.apache.kafka.clients.producer.unknown-metrics-reporter");
         testInvalidInputs("test1,test2");
         testInvalidInputs("org.apache.kafka.common.metrics.FakeMetricsReporter,");
+        testInvalidInputs(TestInterceptorConfig.INTERCEPTOR_CLASSES_CONFIG, TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_CONSUMER_INTERCEPTOR + ", "
+                + TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_CONSUMER_INTERCEPTOR + ", "
+                + TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_CONSUMER_INTERCEPTOR,  org.apache.kafka.test.MockConsumerInterceptor.class);
+        testInvalidInputs(TestInterceptorConfig.INTERCEPTOR_CLASSES_CONFIG, TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_PRODUCER_INTERCEPTOR + ", "
+                + TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_PRODUCER_INTERCEPTOR + ", "
+                + TestInterceptorConfig.ORG_APACHE_KAFKA_TEST_MOCK_PRODUCER_INTERCEPTOR,  org.apache.kafka.test.MockProducerInterceptor.class);
```
We don't have to test both the producer and consumer interceptor classes here; we just have to make sure that after an invocation of `AbstractConfig::getConfiguredInstances`, the right objects are cleaned up. (The value in this test is that we ensure that the cleanup logic is handled directly by the `AbstractConfig` class, instead of being performed out-of-band by, e.g., the `KafkaConsumer` or `KafkaProducer` classes).

In addition, we don't have to use this test case or the pattern of `testValidInputs`/`testInvalidInputs` helper methods.

We can save a lot of complexity by pulling this out into its own test case, and using the existing `TestConfig` class instead of creating a new one:

```java
    @Test
    public void testConfiguredInstancesClosedOnFailure() {
        try {
            Map<String, String> props = new HashMap<>();
            String threeConsumerInterceptors = MockConsumerInterceptor.class.getName() + ", "
                    + MockConsumerInterceptor.class.getName() + ", "
                    + MockConsumerInterceptor.class.getName();
            props.put(TestConfig.METRIC_REPORTER_CLASSES_CONFIG, threeConsumerInterceptors);
            props.put("client.id", "test");
            TestConfig testConfig = new TestConfig(props);

            MockConsumerInterceptor.setThrowOnConfigExceptionThreshold(3);
            assertThrows(
                    Exception.class,
                    () -> testConfig.getConfiguredInstances(TestConfig.METRIC_REPORTER_CLASSES_CONFIG, Object.class)
            );
            assertEquals(3, MockConsumerInterceptor.CONFIG_COUNT.get());
            assertEquals(2, MockConsumerInterceptor.CLOSE_COUNT.get());
        } finally {
            MockConsumerInterceptor.resetCounters();
        }
    }
```

beardt replied:
@C0urante  Done!

beardt replied:
As a follow up on your earlier comments, another value of this test is in understanding the intent of the coded feature.  In this case (pun intended), a developer looking at this test case would rightfully wonder what does `TestConfig.METRIC_REPORT_CLASSES_CONFIG` have to do with interceptors?  

Furthermore, using `TestConfig` in your suggested fashion and goal adds confusion to the test case as `METRIC_REPORT_CLASSES_CONFIG` would never be used with interceptors in the real world.  

**NOTE:**  Although not in scope,  `TestConfig` really should be renamed to `MetricReporterTestConfig` as this more accurately aligns with the implementation.  

Therefore, I recommend retaining the slightly renamed `InterceptorTestConfig` class as this is consistent with the collaborating test subject e.g. `MockConsumerInterceptor` and  reduces noise when reasoning out the test case.   Additionally, I've modified your suggested test case below which includes a slightly updated version of the `InterceptorTestConfig` class your review:   

 ```
@Test
    public void testConfiguredInstancesClosedOnFailure() {
               try {
            Map<String, String> props = new HashMap<>();
            String threeConsumerInterceptors = MockConsumerInterceptor.class.getName() + ", "
                    + MockConsumerInterceptor.class.getName() + ", "
                    + MockConsumerInterceptor.class.getName();
            props.put(InterceptorTestConfig.INTERCEPTOR_CLASSES_CONFIG, threeConsumerInterceptors);
            props.put(InterceptorTestConfig.CLIENT_ID_CONFIG, "test");
        
            InterceptorTestConfig interceptorTestConfig = new InterceptorTestConfig(props);
            MockConsumerInterceptor.setThrowOnConfigExceptionThreshold(interceptorTestConfig.getTargetInterceptor());
            assertThrows(
                    Exception.class,
                    () -> interceptorTestConfig.getConfiguredInstances(InterceptorTestConfig.INTERCEPTOR_CLASSES_CONFIG, Object.class)
            );
            assertEquals(3, MockConsumerInterceptor.CONFIG_COUNT.get());
            assertEquals(3, MockConsumerInterceptor.CLOSE_COUNT.get());
        } finally {
            MockConsumerInterceptor.resetCounters();
        }
    }
```

```
private static class InterceptorTestConfig extends AbstractConfig {
        private final int targetInterceptor = 3;
        private static final ConfigDef CONFIG;
        private static final String INTERCEPTOR_CLASSES_CONFIG_DOC = "A list of classes to use as interceptors.";

        public static final String INTERCEPTOR_CLASSES_CONFIG = "interceptor.classes";
        public static final String CLIENT_ID_CONFIG = "client.id";
        public static final String BOOTSTRAP_SERVERS_CONFIG = "bootstrap.servers";

        static {
            CONFIG = new ConfigDef().define(INTERCEPTOR_CLASSES_CONFIG,
                    Type.LIST,
                    "",
                    Importance.LOW,
                    INTERCEPTOR_CLASSES_CONFIG_DOC);
        }
        public InterceptorTestConfig(Map<?, ?> props) {
            super(CONFIG, props);
        }

        public int getTargetInterceptor() {
            return targetInterceptor;
        }
    }
```

C0urante reviewed (commented):
Thanks @beardt, left a few more comments.

beardt committed:
- Added mayClose to the AbstractConfig getConfiguredInstances and getConfiguredInstance methods
- Added testConfiguredInstancesClosedOnFailure to AbstractConfigTest.  Removed old test and configInterceptorTester

C0urante commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -414,34 +414,51 @@ private <T> T getConfiguredInstance(Object klass, Class<T> t, Map<String, Object
-        return getConfiguredInstance(c, t, originals(configOverrides));
+        try {
+            configuredInstance = getConfiguredInstance(c, t, originals(configOverrides));
+        } catch (Exception e) {
```
If we do the try/catch here, it doesn't work; `configuredInstance` is guaranteed to be null in the catch block.

I was thinking of something like this:
```java
        if (o instanceof Configurable) {
            try {
                ((Configurable) o).configure(configPairs);
            } catch (Exception e) {
                maybeClose(o, "AutoCloseable object constructed and configured during failed call to getConfiguredInstance");
                throw e;
            }
        }
```

being added to [these lines](https://github.com/apache/kafka/blob/7d61d4505a16f09b85f5eb37adeff9c3534ec02c/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L406-L407) in the other variant of `getConfiguredInstance`.

beardt replied:
I went ahead and extended the try/catch to enclose the  `o = Utils.newInstance((String) klass, t);` as well.

beardt committed:
-  Moved try/catch to getConfiguredInstance

beardt committed:
-  Updated unit tests for AbstractConfigTest, KafkaConsumerTest and KafkaProducerTest

beardt committed:
- Extended try/catch to enclose instance creation in getConfiguredInstance

beardt committed:
-- Added InterceptorTestConfig to improve test case clarity of testConfiguredInstancesClosedOnFailure usage of mock interceptors versus mock metric reporters

beardt commented:
I went ahead and reintroduced  the InterceptorTestConfig as I do not think it adds anymore complexity than `TestConfig` and `ClassTestConfig` while also making  `testConfiguredInstancesClosedOnFailure` less confusing due to the usage of `MetricReporters` configs with consumer interceptors.

beardt committed:
-- Added emptyList to InterceptorTestConfig default

beardt commented:
@C0urante  Can we move forward with this PR?

C0urante commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -389,22 +389,27 @@ public void logUnused() {
+            maybeClose(o, "AutoCloseable object constructed and configured during failed call to getConfiguredInstance");
+            throw e;
+        }
 
```
Nit: remove this extra blank line?

beardt replied:
@C0urante: I removed all extra blank lines `getConfiguredInstance`

C0urante commented on clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:
```diff
@@ -599,6 +624,31 @@ public TestConfig(Map<?, ?> props) {
         }
     }
 
+    private static class InterceptorTestConfig extends AbstractConfig {
+        private final int targetInterceptor = 3;
+        private static final ConfigDef CONFIG;
+        private static final String INTERCEPTOR_CLASSES_CONFIG_DOC = "A list of classes to use as interceptors.";
+
+        public static final String INTERCEPTOR_CLASSES_CONFIG = "interceptor.classes";
+        public static final String CLIENT_ID_CONFIG = "client.id";
+        public static final String BOOTSTRAP_SERVERS_CONFIG = "bootstrap.servers";
+
+        static {
+            CONFIG = new ConfigDef().define(INTERCEPTOR_CLASSES_CONFIG,
+                    Type.LIST,
+                    Collections.emptyList(),
+                    Importance.LOW,
+                    INTERCEPTOR_CLASSES_CONFIG_DOC);
+        }
+        public InterceptorTestConfig(Map<?, ?> props) {
+            super(CONFIG, props);
+        }
+
+        public int getTargetInterceptor() {
+            return targetInterceptor;
+        }
+    }
```
Sorry, I'm still not in favor of this additional class. The name of the config property isn't super critical and it's not necessary to try to instantiate metrics reporters when using the `metric.reporters` property in the test config; this is only constructed for testing, after all.

Please remove this class and simplify the testing logic by using the existing `TestConfig` class.

beardt replied:
@C0urante No apologies necessary.  I've reverted the code back to your preference.

C0urante reviewed (commented):
Thanks @beardt, just a couple more thoughts.

beardt committed:
-- Adjusted try/catch for configurable instances.

beardt committed:
-- Revert to using the TestConfig and removed TestInterceptorConfig
-- Removed extra space in getConfiguredInstance

C0urante reviewed (approved):
LGTM, thanks @beardt!

Test failures appear unrelated, will merge and try to backport.
